// @microsoft/tsdoc/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var DocNodeManager_1 = require("./configuration/DocNodeManager");
exports.DocNodeManager = DocNodeManager_1.DocNodeManager;
var TSDocConfiguration_1 = require("./configuration/TSDocConfiguration");
exports.TSDocConfiguration = TSDocConfiguration_1.TSDocConfiguration;
var TSDocTagDefinition_1 = require("./configuration/TSDocTagDefinition");
exports.TSDocTagSyntaxKind = TSDocTagDefinition_1.TSDocTagSyntaxKind;
exports.TSDocTagDefinition = TSDocTagDefinition_1.TSDocTagDefinition;
var TSDocValidationConfiguration_1 = require("./configuration/TSDocValidationConfiguration");
exports.TSDocValidationConfiguration = TSDocValidationConfiguration_1.TSDocValidationConfiguration;
var StandardTags_1 = require("./details/StandardTags");
exports.StandardTags = StandardTags_1.StandardTags;
var StandardModifierTagSet_1 = require("./details/StandardModifierTagSet");
exports.StandardModifierTagSet = StandardModifierTagSet_1.StandardModifierTagSet;
var ModifierTagSet_1 = require("./details/ModifierTagSet");
exports.ModifierTagSet = ModifierTagSet_1.ModifierTagSet;
var PlainTextEmitter_1 = require("./emitters/PlainTextEmitter");
exports.PlainTextEmitter = PlainTextEmitter_1.PlainTextEmitter;
var StringBuilder_1 = require("./emitters/StringBuilder");
exports.StringBuilder = StringBuilder_1.StringBuilder;
var TSDocEmitter_1 = require("./emitters/TSDocEmitter");
exports.TSDocEmitter = TSDocEmitter_1.TSDocEmitter;
__export(require("./nodes"));
var ParserContext_1 = require("./parser/ParserContext");
exports.ParserContext = ParserContext_1.ParserContext;
var ParserMessage_1 = require("./parser/ParserMessage");
exports.ParserMessage = ParserMessage_1.ParserMessage;
var ParserMessageLog_1 = require("./parser/ParserMessageLog");
exports.ParserMessageLog = ParserMessageLog_1.ParserMessageLog;
var TextRange_1 = require("./parser/TextRange");
exports.TextRange = TextRange_1.TextRange;
var Token_1 = require("./parser/Token");
exports.Token = Token_1.Token;
exports.TokenKind = Token_1.TokenKind;
var TokenSequence_1 = require("./parser/TokenSequence");
exports.TokenSequence = TokenSequence_1.TokenSequence;
var TSDocParser_1 = require("./parser/TSDocParser");
exports.TSDocParser = TSDocParser_1.TSDocParser;
var DocNodeTransforms_1 = require("./transforms/DocNodeTransforms");
exports.DocNodeTransforms = DocNodeTransforms_1.DocNodeTransforms;
