// @rushstack/ts-command-line/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * An object-oriented command-line parser for TypeScript projects.
 *
 * @packageDocumentation
 */
var CommandLineAction_1 = require("./providers/CommandLineAction");
exports.CommandLineAction = CommandLineAction_1.CommandLineAction;
var BaseClasses_1 = require("./parameters/BaseClasses");
exports.CommandLineParameterKind = BaseClasses_1.CommandLineParameterKind;
exports.CommandLineParameter = BaseClasses_1.CommandLineParameter;
exports.CommandLineParameterWithArgument = BaseClasses_1.CommandLineParameterWithArgument;
var CommandLineFlagParameter_1 = require("./parameters/CommandLineFlagParameter");
exports.CommandLineFlagParameter = CommandLineFlagParameter_1.CommandLineFlagParameter;
var CommandLineStringParameter_1 = require("./parameters/CommandLineStringParameter");
exports.CommandLineStringParameter = CommandLineStringParameter_1.CommandLineStringParameter;
var CommandLineStringListParameter_1 = require("./parameters/CommandLineStringListParameter");
exports.CommandLineStringListParameter = CommandLineStringListParameter_1.CommandLineStringListParameter;
var CommandLineIntegerParameter_1 = require("./parameters/CommandLineIntegerParameter");
exports.CommandLineIntegerParameter = CommandLineIntegerParameter_1.CommandLineIntegerParameter;
var CommandLineChoiceParameter_1 = require("./parameters/CommandLineChoiceParameter");
exports.CommandLineChoiceParameter = CommandLineChoiceParameter_1.CommandLineChoiceParameter;
var CommandLineRemainder_1 = require("./parameters/CommandLineRemainder");
exports.CommandLineRemainder = CommandLineRemainder_1.CommandLineRemainder;
var CommandLineParameterProvider_1 = require("./providers/CommandLineParameterProvider");
exports.CommandLineParameterProvider = CommandLineParameterProvider_1.CommandLineParameterProvider;
var CommandLineParser_1 = require("./providers/CommandLineParser");
exports.CommandLineParser = CommandLineParser_1.CommandLineParser;
var DynamicCommandLineAction_1 = require("./providers/DynamicCommandLineAction");
exports.DynamicCommandLineAction = DynamicCommandLineAction_1.DynamicCommandLineAction;
var DynamicCommandLineParser_1 = require("./providers/DynamicCommandLineParser");
exports.DynamicCommandLineParser = DynamicCommandLineParser_1.DynamicCommandLineParser;
var CommandLineHelper_1 = require("./CommandLineHelper");
exports.CommandLineHelper = CommandLineHelper_1.CommandLineHelper;
